flowmachine/tests/test_cache_utils.py

```diff
@@ -0,0 +1,182 @@
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
+Tests for cache management utilities.
+"""
+from unittest.mock import Mock
+
+import pytest
+
+from flowmachine.core.cache import (
+    rescore,
+    compute_time,
+    shrink_below_size,
+    shrink_one,
+    size_of_cache,
+    size_of_table,
+    score,
+    get_query_by_id,
+)
+from flowmachine.features import daily_location
+
+
+def test_rescore():
+    """Test that score updating algorithm is correct"""
+    connection_mock = Mock()
+    connection_mock.fetch.side_effect = [
+        [[10]],
+        [[5]],
+        [[1]],
+    ]  # Byte size, runtime, last score
+    query_mock = Mock()
+    query_mock.table_name = "DUMMY_SCHEMA.DUMMY_NAME"
+    new_score = rescore(connection_mock, query_mock, half_life=2)
+    assert 1 + (5 / 1000) / 10 * (1 + 2) == new_score
+
+
+def test_compute_time():
+    """Compute time should take value returned in ms and turn it into seconds."""
+    connection_mock = Mock()
+    connection_mock.fetch.return_value = [[10]]
+    assert 10 / 1000 == compute_time(connection_mock, "DUMMY_ID")
+
+
+def test_shrink_one(flowmachine_connect):
+    """Test that shrink_one removes a cache record."""
+    dl = daily_location("2016-01-01").store().result()
+    dl_aggregate = dl.aggregate().store().result()
+    flowmachine_connect.engine.execute(
+        f"UPDATE cache.cached SET cache_score = 1 WHERE query_id='{dl_aggregate.md5}'"
+    )
+    flowmachine_connect.engine.execute(
+        f"UPDATE cache.cached SET cache_score = 0.5 WHERE query_id='{dl.md5}'"
+    )
+    removed_query, table_size = shrink_one(flowmachine_connect)
+    assert dl.md5 == removed_query.md5
+    assert not dl.is_stored
+    assert dl_aggregate.is_stored
+
+
+def test_shrink_to_size_does_nothing_when_cache_ok(flowmachine_connect):
+    """Test that shrink_below_size doesn't remove anything if cache size is within limit."""
+    dl = daily_location("2016-01-01").store().result()
+    removed_queries = shrink_below_size(
+        flowmachine_connect, size_of_cache(flowmachine_connect)
+    )
+    assert 0 == len(removed_queries)
+    assert dl.is_stored
+
+
+def test_shrink_to_size_removes_queries(flowmachine_connect):
+    """Test that shrink_below_size removes queries when cache limit is breached."""
+    dl = daily_location("2016-01-01").store().result()
+    removed_queries = shrink_below_size(
+        flowmachine_connect, size_of_cache(flowmachine_connect) - 1
+    )
+    assert 1 == len(removed_queries)
+    assert not dl.is_stored
+
+
+def test_shrink_to_size_respects_dry_run(flowmachine_connect):
+    """Test that shrink_below_size doesn't remove anything during a dry run."""
+    dl = daily_location("2016-01-01").store().result()
+    removed_queries = shrink_below_size(
+        flowmachine_connect, size_of_cache(flowmachine_connect) - 1, dry_run=True
+    )
+    assert 1 == len(removed_queries)
```

Comment:
It would be useful to have this test for a case where `len(removed_queries) > 1` (see earlier comment in `cache.py`)

Reply:
@jch1g10 this should be sorted now.